File: ts/src/day13.ts
import { DayBase } from './day-runner.js';

const foldText = "fold along ";
const foldTextLen = foldText.length;

export class Day13 extends DayBase {
    constructor() {
        super(13);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        // const lines = [
        //     '6,10', '0,14', '9,10', '0,3', '10,4', '4,11', '6,0', '6,12',
        //     '4,1', '0,13', '10,12', '3,4', '3,0', '8,4', '1,10', '2,14', '8,10', '9,0',

        //     'fold along y=7', 'fold along x=5',
        // ];

        let dots: number[][] = []
        let folds: number[][] = [];
        lines.forEach(line => {
            if (line.startsWith("fold")) {
                const fold = line.substring(foldTextLen).trim().split('=');
                const v = parseInt(fold[1], 10);
                folds.push(fold[0] === 'x' ? [v, 0] : [0, v]);
            } else if (line) {
                const dot = line.trim().split(',');
                dots.push([parseInt(dot[0], 10), parseInt(dot[1], 10)]);
            }
        });

        const part1 = this.fold(dots, folds[0]).length;
        let part2: number[][] = [...dots];
        for (let fold of folds) {
            part2 = this.fold(part2, fold);
        }

        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, part1);
        console.info('  Part 2:');
        this.draw(part2);
        console.info('');
    }

    private fold(dots: number[][], fold: number[]): number[][] {
        let result: number[][] = [];
        const folded = (fv: number, v: number): number => v > fv ? fv - (v - fv) : v;
        dots.forEach(dot => {
            let d: number[];
            if (fold[0] === 0) { // fold along y=?. x stays the same
                d = [dot[0], folded(fold[1], dot[1])];
            } else {             // fold along x=?. y stays the same
                d = [folded(fold[0], dot[0]), dot[1]];
            }
            if (!this.includes(result, d)) {
                result.push(d);
            }
        })
        return result;
    }

    private includes(dots: number[][], dot: number[]): boolean {
        for (let d of dots) {
            if (d[0] == dot[0] && d[1] == dot[1]) {
                return true;
            }
        }
        return false;
    }

    private draw(dots: number[][]) {
        let minx = 1000000, maxx = 0, miny = 1000000, maxy = 0;
        for (let dot of dots) {
            minx = dot[0] < minx ? dot[0] : minx;
            maxx = dot[0] > maxx ? dot[0] : maxx;
            miny = dot[1] < miny ? dot[1] : miny;
            maxy = dot[1] > maxy ? dot[1] : maxy;
        }
        //console.info(`X: ${minx}-${maxx}; Y: ${miny}-${maxy}`);

        for (let y = miny; y <= maxy; y++) {
            let line = '';
            for (let x = minx; x <= maxx; x++) {
                line += this.includes(dots, [x,y]) ? '#' : ' ';
            }
            console.info('    ' + line);
        }
    }
}

File: py/day13.py
import os

# https://adventofcode.com/2021/day/13

def foldValue(foldVal, val):
    return val if val < foldVal else foldVal - (val - foldVal)

def fold(dots, fold):
    result = {}
    for dot in dots:
        if fold[0] == 0: # fold along y=?, x stays the same
            result[(dot[0], foldValue(fold[1], dot[1]))] = True
        else:            # fold along x=?, y stays the same
            result[(foldValue(fold[0], dot[0]), dot[1])] = True
    return result

def draw(dots):
    xsort = sorted(dots, key = lambda k: k[0])
    minx = xsort[0][0]
    maxx = xsort[-1][0]
    ysort = sorted(dots, key = lambda k: k[1])
    miny = ysort[0][1]
    maxy = ysort[-1][1]
    for y in range(miny, maxy + 1):
        line = ''
        for x in range(minx, maxx + 1):
            line += '#' if (x,y) in dots else ' '
        print('    ' + line)

foldTextLen = len('fold along ')
folds = []
dots = {}
with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day13-input", "r") as f:
    for line in [line.strip() for line in f.readlines()]:
        if line.startswith("fold"):
            f = line[foldTextLen:].split('=')
            v = int(f[1])
            folds.append([v, 0] if f[0] == 'x' else [0, v])
        elif line:
            dot = (int(line.split(',')[0]), int(line.split(',')[1]))
            dots[dot] = True

print("Day 13")
print("  Part 1", len(fold(dots, folds[0])))

part2 = dots
for f in folds:
    part2 = fold(part2, f)

print("  Part 2")
draw(part2)

